node/lib/util/sequencer_state_util.js

```diff
@@ -0,0 +1,237 @@
+    assert.isString(gitDir);
+    const content = yield exports.readFile(gitDir, COMMITS_FILE);
+    if (null !== content) {
+        const lines = content.split("\n");
+        const nonEmpty = lines.filter(line => line.length > 0);
+        if (0 !== nonEmpty.length) {
+            return nonEmpty;
+        }
+    }
+    return null;
+});
+
+/**
+ * Return the index of the current commit stored in the specified `gitDir`
+ * having the specified `numCommits` number of commits  or null if the file is
+ * missing, malformed, our out-of-bounds.
+ *
+ * @param {String} gitDir
+ * @param {Number} numCommits
+ * @return {Number}
+ */
+exports.readCurrentCommit = co.wrap(function *(gitDir, numCommits) {
+    assert.isString(gitDir);
+    assert.isNumber(numCommits);
+    const content = yield exports.readFile(gitDir, CURRENT_COMMIT_FILE);
+    if (null !== content) {
+        const lines = content.split("\n");
+        if (0 !== lines.length) {
+            const index = Number.parseInt(lines[0]);
+            if (!Number.isNaN(index) && index < numCommits) {
+                return index;
+            }
+        }
+    }
+    return null;
+});
+
+/**
+ * Return the sequencer state if it exists in the specified `gitDir`, or null
+ * if it is missing or malformed.
+ * TODO: emit diagnostic when malformed?
+ *
+ * @param {String} gitDir
+ * @return {SequencerState|null}
+ */
+exports.readSequencerState = co.wrap(function *(gitDir) {
+    assert.isString(gitDir);
+
+    const typeContent = yield exports.readFile(gitDir, TYPE_FILE);
+    if (null === typeContent) {
+        return null;                                                  // RETURN
+    }
+    const typeLines = typeContent.split("\n");
+    if (2 !== typeLines.length) {
+        return null;                                                  // RETURN
+    }
+    const type = typeLines[0];
+    if (!(type in SequencerState.TYPE)) {
+        return null;                                                  // RETURN
+    }
+    const original = yield exports.readCommitAndRef(gitDir,
+                                                    ORIGINAL_HEAD_FILE);
+    if (null === original) {
+        return null;                                                  // RETURN
+    }
+    const target = yield exports.readCommitAndRef(gitDir, TARGET_FILE);
+    if (null === target) {
+        return null;                                                  // RETURN
+    }
+    const commits = yield exports.readCommits(gitDir);
+    if (null === commits) {
+        return null;                                                  // RETURN
+    }
+    const currentCommit = yield exports.readCurrentCommit(gitDir,
+                                                          commits.length);
+    if (null === currentCommit) {
+        return null;                                                  // RETURN
+    }
+    return new SequencerState(type, original, target, commits, currentCommit);
```

Comment:
Would a fluent builder pattern be better here, or is that not a thing in JS?

Reply:
To deal reduce the number of args? I think what you'd be more likely to see in JS would be a dictionary, especially if you have a optional arguments. I didn't use one because all the args here are required, but it would look like:
```JavaScript
SequencerState({
    type: type
    original: original,
    target: target,
    commits: commits,
    currentCommit: currentCommit,
});
```

Reply:
Not just reduce the number, but also avoid having to look up ordering.  I like that version better.

Reply:
I'm fine with that. I could see this constructor getting more args in the future, and I don't think I'd go to 6.